code/modules/xenoarchaeology/xenoartifact.dm

```diff
@@ -0,0 +1,470 @@
+
+	if(prob(30) || icon_slots[2])//Bottom
+		if(!(icon_slots[2]))
+			icon_slots[2] = rand(1, 2)
+		generate_icon(icon, "IBTM[icon_slots[2]]")
+		generate_icon(icon, "IBTML[icon_slots[2]]", material)
+
+/obj/item/xenoartifact/CanAllowThrough(atom/movable/mover, turf/target) //tweedle dee, density feature
+	if(get_trait(/datum/xenoartifact_trait/minor/dense))
+		return FALSE
+	return ..()
+
+/obj/item/xenoartifact/attack_hand(mob/user) //tweedle dum, density feature
+	if(get_trait(/datum/xenoartifact_trait/minor/dense))
+		if(process_type == PROCESS_TYPE_LIT) //Snuff out candle
+			to_chat(user, "<span class='notice'>You snuff out [name]</span>")
+			process_type = null
+			return FALSE
+			
+		if(user.a_intent != INTENT_GRAB)
+			SEND_SIGNAL(src, XENOA_INTERACT, null, user, user) //Calling the regular attack_hand signal causes feature issues, like picking up the artifact.
+		else if(touch_desc?.on_touch(src, user) && user.can_see_reagents())
+			balloon_alert(user, (initial(touch_desc.desc) ? initial(touch_desc.desc) : initial(touch_desc.label_name)), material)
+		return FALSE
+	..()
+
+/obj/item/xenoartifact/examine(mob/living/carbon/user)
+	. = ..()	
+	if(isobserver(user))
+		to_chat(user, "<span class='notice'>[special_desc]</span>")
+		for(var/datum/xenoartifact_trait/t as() in traits)
+			to_chat(user, "<span class='notice'>[t?.desc ? t.desc : t.label_name]</span>")
+			to_chat("test")
+	else if(iscarbon(user) && user.can_see_reagents()) //Not checking carbon throws a runtime concerning observers
+		to_chat(user, "<span class='notice'>[special_desc]</span>")
+	return (label_desc ? . + label_desc : .)
+
+/obj/item/xenoartifact/interact(mob/user)
+	. = ..()
+	if(process_type == PROCESS_TYPE_LIT) //Snuff out candle
+		to_chat(user, "<span class='notice'>You snuff out [name]</span>")
+		process_type = null
+		return
+	if(user.a_intent == INTENT_GRAB)
+		if(touch_desc?.on_touch(src, user) && user.can_see_reagents())
+			balloon_alert(user, (initial(touch_desc.desc) ? initial(touch_desc.desc) : initial(touch_desc.label_name)), material)
+		return
+	SEND_SIGNAL(src, XENOA_INTERACT, null, user, user)
+
+/obj/item/xenoartifact/attackby(obj/item/I, mob/living/user, params)
+	var/tool_text
+	for(var/datum/xenoartifact_trait/t as() in traits) //chat & bubble hints & helpers
+		if(t?.on_item(src, user, I) && user.can_see_reagents())
+			tool_text = "[tool_text][t.desc ? t.desc : t.label_name]\n"
+	if(tool_text)
+		balloon_alert(user, tool_text, material)
+
+	//abort if grab intent
+	if(!(COOLDOWN_FINISHED(src, xenoa_cooldown))||user?.a_intent == INTENT_GRAB||istype(I, /obj/item/xenoartifact_label)||istype(I, /obj/item/xenoartifact_labeler))
+		if(user?.a_intent == INTENT_GRAB)
+			to_chat(user, "<span class='notice'>You preform a safe operation on [src] with [I].</span>")
+		return
+	else if(istype(I, /obj/item/wirecutters) && (locate(/obj/item/xenoartifact_label) in contents)) //allow people to remove stickers
```

Comment:
Should also probably do use_tool somewhere, since I think that can perform some special checks and play the sound. Have a look and see how it is done elsewhere.